### PR TITLE
test: Refactor browser and testdapp page objects

### PR DESCRIPTION
**Pull request by cortisiko**
## **Description**

This pull request focuses on refactoring the browser and testDapp page objects to improve code readability and facilitate debugging. A recent update in Detox enables webview support on iOS, which was previously only available on Android. This upgrade was integrated into this pull request and greatly simplifies the browser-related tests. Additionally, we will have the ability to consistently inject code into the web view for testing (This is especially important for any sort of Dapp-API tests that we intend on writing in the future).  

But why upgrade Detox specifically in this PR? As part of the browser refactor, we want to interact with elements consistently, which reduces the possibility of test flakiness. For instance, we used xy coordinates to interact with elements in the web view, which posed challenges. However, these tests would break if the device screen size changed, necessitating the acquisition of new xy coordinates. By upgrading Detox to version 20.20, we eliminate this issue and reduce the likelihood of encountering unreliable tests when running them on various devices. 

 You can read up more on the different web view strategies and web actions here: https://wix.github.io/Detox/docs/api/webviews 

Things to be aware of:

- The matchers class has been updated with new web view locator strategies. These strategies now include: byID, ByCssSelector, ByClassName, and ByxPath, providing more flexibility and options for locating elements in the web view. 

- jsDocs was added for the new strategies 

- All of the confirmation tests except: send-eth.spec.js and advanced-gas-fees.spec.js were updated. 

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: .detoxrc.js [hunk 1]
```diff
@@ -68,7 +68,7 @@ module.exports = {
     'android.emulator': {
       type: 'android.emulator',
       device: {
-        avdName: 'Pixel_5_API_30',
+        avdName: 'Pixel_3_API_34',
```
Comment (cortisiko) [inline .detoxrc.js:71]: note to self: revert this change

File: e2e/selectors/Browser/TestDapp.selectors.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// eslint-disable-next-line import/prefer-default-export
+export const TestDappSelectorsWebIDs = {
```
Comment (cortisiko) [inline e2e/selectors/Browser/TestDapp.selectors.js:2]: test dapp selector id's

File: e2e/specs/confirmations/approve-custom-erc20.spec.js [hunk 3]
```diff
@@ -69,10 +67,9 @@ describe(SmokeConfirmations('ERC20 tokens'), () => {
 
         // Navigate to the activity screen
         await TabBarComponent.tapActivity();
-
         // Assert erc20 is approved
-        await TestHelpers.checkIfElementByTextIsVisible(
-          enContent.transaction.confirmed,
+        await Assertions.checkIfTextIsDisplayed(
```
Comment (cortisiko) [inline e2e/specs/confirmations/approve-custom-erc20.spec.js:71]: cleaning up assertion

File: e2e/specs/confirmations/approve-default-erc20.spec.js [hunk 4]
```diff
@@ -44,42 +47,47 @@ describe(SmokeConfirmations('ERC20 tokens'), () => {
...
+        //   ),
+        // ).toHaveText(EXPECTED_TOKEN_AMOUNT);
+        await Assertions.checkIfElementToHaveText(
+          ContractApprovalModal.approveTokenAmount,
+          EXPECTED_TOKEN_AMOUNT,
+        );
         // Tap next button
-        await TestHelpers.checkIfElementWithTextIsVisible(
-          enContent.transaction.next,
+        await Assertions.checkIfTextIsDisplayed(
```
Comment (cortisiko) [inline e2e/specs/confirmations/approve-default-erc20.spec.js:64]: cleaning up assertion

File: e2e/specs/confirmations/approve-default-erc20.spec.js [hunk 5]
```diff
@@ -44,42 +47,47 @@ describe(SmokeConfirmations('ERC20 tokens'), () => {
...
+          ContractApprovalModalSelectorsText.NEXT,
         );
-        await TestHelpers.tapByText(enContent.transaction.next);
+        await ContractApprovalModal.tapNextButton();
 
-        // Tap approve button
-        await TestHelpers.checkIfElementWithTextIsVisible(
-          enContent.transactions.tx_review_approve,
+        await Assertions.checkIfTextIsDisplayed(
```
Comment (cortisiko) [inline e2e/specs/confirmations/approve-default-erc20.spec.js:69]: cleaning up assertion

File: e2e/specs/confirmations/approve-default-erc20.spec.js [hunk 6]
```diff
@@ -44,42 +47,47 @@ describe(SmokeConfirmations('ERC20 tokens'), () => {
...
+        await ContractApprovalModal.tapApproveButton();
 
         // Navigate to the activity screen
         await TabBarComponent.tapActivity();
 
         // Assert erc20 is approved
-        await TestHelpers.checkIfElementByTextIsVisible(
-          enContent.transaction.confirmed,
+
+        await Assertions.checkIfTextIsDisplayed(
```
Comment (cortisiko) [inline e2e/specs/confirmations/approve-default-erc20.spec.js:80]: cleaning up assertion

File: e2e/specs/confirmations/send-erc20-with-dapp.spec.js [hunk 7]
```diff
@@ -42,23 +42,24 @@ describe(SmokeConfirmations('ERC20 tokens'), () => {
...
         // Tap confirm button
         await TestDApp.tapConfirmButton();
 
         // Navigate to the activity screen
         await TabBarComponent.tapActivity();
 
         // Assert "Sent Tokens" transaction is displayed
-        await TestHelpers.checkIfElementByTextIsVisible(
-          SENT_TOKENS_MESSAGE_TEXT,
+        await Assertions.checkIfTextIsDisplayed(
```
Comment (cortisiko) [inline e2e/specs/confirmations/send-erc20-with-dapp.spec.js:61]: cleaning up assertion

File: e2e/selectors/Browser/BrowserURLBar.selectors.js [hunk 8]
```diff
@@ -0,0 +1,6 @@
+// eslint-disable-next-line import/prefer-default-export
+export const BrowserURLBarSelectorsIDs = {
```
Comment (cortisiko) [inline e2e/selectors/Browser/BrowserURLBar.selectors.js:2]: Added Selector object for BrowserURLBar

File: e2e/selectors/Browser/AddFavorites.selectors.js [hunk 9]
```diff
@@ -0,0 +1,8 @@
+// eslint-disable-next-line import/prefer-default-export
```
Comment (cortisiko) [inline e2e/selectors/Browser/AddFavorites.selectors.js:1]: Added Selector object for Add favorites view

File: e2e/pages/Browser/TestDApp.js [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+  async tapNFTTransferButton() {
+    await this.tapButton(this.nftTransferFromTokensButton);
+  }
+
+  async tapConfirmButton() {
+    await Gestures.tap(this.confirmButtonText, 0);
+  }
+
+  async tapButton(elementId) {
+    await Gestures.scrollToWebViewPort(elementId);
```
Comment (cortisiko) [inline e2e/pages/Browser/TestDApp.js:122]: Before tapping on an element in the test dapp, we are going to scroll to that element

File: e2e/pages/Browser/TestDApp.js [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    await Gestures.tapWebElement(elementId);
+  }
+
+  async navigateToTestDappWithContract({ contractAddress }) {
+    await Browser.tapUrlInputBox();
+    await Browser.navigateToURL(
+      `${TEST_DAPP_LOCAL_URL}?scrollTo=''&contract=${contractAddress}`,
```
Comment (cortisiko) [inline e2e/pages/Browser/TestDApp.js:129]: This one is a bit interesting. If we were to exclude `?scrollTo=` in the query param, the dapp method that we are trying to tap on would not be active. `?scrollTo=` is needed

File: e2e/utils/Gestures.js [hunk 12]
```diff
@@ -155,6 +157,15 @@ class Gestures {
       .swipe(direction, speed, percentage, xStart, yStart);
   }
 
+  /**
+   * Scrolls the web element until its top is at the top of the viewport.
+   * @param {Promise<Element>} elementID - A promise resolving to the target element's ID or element itself.
+   */
+  static async scrollToWebViewPort(elementID) {
```
Comment (cortisiko) [inline e2e/utils/Gestures.js:164]: added a user action to scroll to a specific element in the webview

File: e2e/utils/Matchers.js [hunk 13]
```diff
@@ -76,31 +76,46 @@ class Matchers {
...
    * Get element by web ID.
    *
-   * @param {string} webID - The web ID of the element to locate
+   * * @param {string} webviewID - The web ID of the inner element to locate within the webview
+   *  @param {string} innerID - The web ID of the browser webview
    * @return {Promise<Detox.IndexableWebElement>} Resolves to the located element
    */
-  static async getElementByWebID(webID) {
-    return web.element(by.web.id(webID));
+  static async getElementByWebID(webviewID, innerID) {
```
Comment (cortisiko) [inline e2e/utils/Matchers.js:83]: Strategy to locate element by webID

File: e2e/utils/Matchers.js [hunk 14]
```diff
@@ -76,31 +76,46 @@ class Matchers {
...
    * Get element by CSS selector.
-   *
+   * @param {string} webviewID - The web ID of the browser webview
    * @param {string} selector - CSS selector to locate the element
    * @return {Promise<Detox.IndexableWebElement>} - Resolves to the located element
    */
-  static async getElementByCSSSelector(selector) {
-    return web.element(by.web.cssSelector(selector));
+
+  static async getElementByCSS(webviewID, selector) {
```
Comment (cortisiko) [inline e2e/utils/Matchers.js:95]: Strategy to locate element by CSSSelector

File: e2e/utils/Matchers.js [hunk 15]
```diff
@@ -76,31 +76,46 @@ class Matchers {
...
   }
 
   /**
    * Get element by XPath.
-   *
+   * @param {string} webviewID - The web ID of the browser webview
+   * @param {string} xpath - XPath expression to locate the element
+   * @return {Promise<Detox.IndexableWebElement>} - Resolves to the located element
+   */
+  static async getElementByXPath(webviewID, xpath) {
```
Comment (cortisiko) [inline e2e/utils/Matchers.js:106]: Strategy to locate element by xPath

File: e2e/utils/Matchers.js [hunk 16]
```diff
@@ -76,31 +76,46 @@ class Matchers {
...
+  }
+  /**
+   * Get element by href.
+   * @param {string} webviewID - The web ID of the browser webview
    * @param {string} xpath - XPath expression to locate the element
    * @return {Promise<Detox.IndexableWebElement>} - Resolves to the located element
    */
-  static async getElementByXPath(xpath) {
-    return web.element(by.web.xpath(xpath));
+  static async getElementByHref(webviewID, url) {
```
Comment (cortisiko) [inline e2e/utils/Matchers.js:116]: Strategy to locate element by href

**Review by cortisiko — COMMENTED** [review]
self review: left comments to clarify what was done

File: e2e/utils/Gestures.js (hunk 12 above)
Comment (chrisleewilcox) [reply e2e/utils/Gestures.js:166]: Since we are passing elements that may have been created using text or other selector, wouldn't this parameter be something like `elem` instead of `elementID`?

**Review by Andepande — DISMISSED** [review]
LGTM Dev approval remaining and update with main

File: e2e/utils/Gestures.js (hunk 12 above)
Comment (cortisiko) [reply e2e/utils/Gestures.js]: fair point, I have updated the code.

**Review by georgewrmarshall — DISMISSED** [review]
LGTM! Approving changes to the component-library folder to unblock as code owner for @MetaMask/design-system

**Review by sleepytanya — DISMISSED** [review]
Looks great!

File: .detoxrc.js (hunk 1 above)
Comment (Cal-L) [reply .detoxrc.js:71]: Does this still need to be reverted before merge?

**Review by Cal-L — COMMENTED** [review]
Overall, looks pretty good. Just left one question

File: .detoxrc.js (hunk 1 above)
Comment (cortisiko) [reply .detoxrc.js]: thanks for pointing this out! I undo this change

**Review by Cal-L — DISMISSED** [review]
LGTM

**Review by chrisleewilcox — DISMISSED** [review]
LGTM

**Review by Andepande — APPROVED** [review]
LG
